### PR TITLE
ARROW-1342: [Python] Support strided ndarrays in pandas conversion from nested lists

### PR DESCRIPTION
**Pull request by wesm**
This does drop the vector append to the builder that was there before. I'm going to do some local benchmarking to make sure this doesn't degrade performance unacceptably, will report back here

### CODE REVIEW

**Comment by wesm** [comment]
The performance doesn't appear to be meaningfully different:

setup code

```
from collections import OrderedDict
import pyarrow as pa
import pandas as pd
import numpy as np
include_index = False

arrays = OrderedDict()
fields = []

fields.append(pa.field('int64', pa.list_(pa.int64())))
arrays['int64'] = [
    np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
    np.array([0, 1, 2, 3, 4]),
    None,
    np.array([0])
]
fields.append(pa.field('double', pa.list_(pa.float64())))
arrays['double'] = [
    np.array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]),
    np.array([0., 1., 2., 3., 4.]),
    None,
    np.array([0.])
]
fields.append(pa.field('str_list', pa.list_(pa.string())))
arrays['str_list'] = [
    np.array([u"1", u"ä"]),
    None,
    np.array([u"1"]),
    np.array([u"1", u"2", u"3"])
]

if include_index:
    fields.append(pa.field('__index_level_0__', pa.int64()))
df = pd.DataFrame(arrays)

df = pd.concat([df] * 10000, ignore_index=True)
```

before:

```
In [12]: timeit pa.Array.from_pandas(df['int64'])
13.6 ms ± 492 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

after:

```
In [2]: timeit pa.Array.from_pandas(df['int64'])
13.1 ms ± 729 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by wesm** [comment]
Updated the benchmarks, was accidentally picking up a debug library

File: cpp/src/arrow/python/pandas_to_arrow.cc [hunk 1]
```diff
@@ -125,37 +123,42 @@ static int64_t MaskToBitmap(PyArrayObject* mask, int64_t length, uint8_t* bitmap
...
     return Status::Invalid("only handle 1-dimensional arrays");
   }
 
-  if (PyArray_DESCR(numpy_array)->type_num != np_type) {
-    return Status::Invalid("can only handle exact conversions");
+  const int received_type = PyArray_DESCR(numpy_array)->type_num;
+  if (received_type != np_type) {
+    std::stringstream ss;
+    ss << "trying to convert NumPy type " << GetNumPyTypeName(np_type) << " but got "
+       << GetNumPyTypeName(received_type);
```
Comment (xhochy) [inline cpp/src/arrow/python/pandas_to_arrow.cc]: +1, I like these more informative error messages

**Review by xhochy — APPROVED** [review]

**Comment by wesm** [comment]
Fixing the build, then going to merge this and cut an RC

**Pull request closed without merging**